File: frontend/src/services/programService.ts
import { api } from '@/lib/api'
import type { Program, ProgramApplication, PaginatedResponse } from '@/types'

export const programService = {
  list: (category?: string) => api.get<PaginatedResponse<Program>>('/programs', { params: { category } }),
  get: (id: number) => api.get<{ data: Program }>(`/programs/${id}`),
  myApplications: () => api.get<PaginatedResponse<ProgramApplication>>('/farmer/program-applications'),
  apply: (programId: number, formData: FormData) =>
    api.post<{ message: string; data: ProgramApplication }>(`/farmer/programs/${programId}/apply`, formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    }),
}

File: mobile/lib/services/program_service.dart
import 'package:dio/dio.dart';

import '../core/api/api_client.dart';
import '../models/program.dart';

class ProgramService {
  ProgramService(this._api);

  final ApiClient _api;

  Future<List<Program>> list() {
    return _api.handle(
      _api.get('/programs'),
      (json) {
        final data = (json as Map<String, dynamic>)['data'] as List<dynamic>;
        return data.map((e) => Program.fromJson(e as Map<String, dynamic>)).toList();
      },
    );
  }

  Future<List<ProgramApplication>> myApplications() {
    return _api.handle(
      _api.get('/farmer/program-applications'),
      (json) {
        final data = (json as Map<String, dynamic>)['data'] as List<dynamic>;
        return data.map((e) => ProgramApplication.fromJson(e as Map<String, dynamic>)).toList();
      },
    );
  }

  Future<String> apply(int programId, {String? remarks, List<String>? documentPaths}) async {
    final form = FormData.fromMap({
      if (remarks != null && remarks.isNotEmpty) 'remarks': remarks,
    });
    if (documentPaths != null) {
      for (final path in documentPaths) {
        form.files.add(MapEntry('documents[]', await MultipartFile.fromFile(path)));
      }
    }
    return _api.handle(
      _api.postMultipart('/farmer/programs/$programId/apply', form),
      (json) => (json as Map<String, dynamic>)['message'] as String? ?? 'Application submitted.',
    );
  }
}
